### ISSUE TITLE
Sniff-specific ignore rules inside rulesets are filtering out too many files

### ISSUE THREAD

**Issue by wvega** [squizlabs/PHP_CodeSniffer#2391]
Consider the following custom standard (`phpcs.xml`):

```xml
<?xml version="1.0"?>
<ruleset name="Custom Standard">
    <rule ref="WordPress">
        <exclude-pattern>/the/content/of/the/pattern/is/irrelevant/</exclude-pattern>
    </rule>
</ruleset>
```

If the path to the project using that standard file includes `wordpress` (the name of the referenced rule) at any level, PHPCS will ignore all files on that project. The list of files to check is empty, even if other rules are being referenced in the standard without using an exclude pattern.

The way I understand it, on v3.4.0 and up to https://github.com/squizlabs/PHP_CodeSniffer/commit/8c94cf9231dc70e8cb45c52a8f3a34501e271da0 (at least), when exclude-patterns are defined for specific standards, categories or sniffs, the `Ruleset::getIgnorePatterns()` method returns an array with a structure similar to:

```php
$ignorePatterns = [
    'WordPress' => [
        '/the/content/of/the/pattern/is/irrelevant/' => 'absolute'
    ],
    ...
    'WordPress.PHP.YodaConditions.NotYoda' => [
        '/the/content/of/the/pattern/is/irrelevant/' => 'absolute'
    ],
    ...
]
```

That array is later used on the [shouldIgnorePath() method](https://github.com/squizlabs/PHP_CodeSniffer/blob/master/src/Filters/Filter.php#L231-L263), as follows:

```php
foreach ($ignorePatterns as $pattern => $type) {
    // ...
    if (preg_match($pattern, $testPath) === 1) {
        return true;
    }
}
```

So that the name of the standard, category or sniff ends up being used as the pattern to ignore. 

When the exclude-pattern is specific to a category of sniffs or a single sniff, then is very unlikely that the name of that category or sniff matches the path of any file on the project, but when the exclude-pattern is specific to a standard with a single word name (like `WordPress` above) the problem is more likely to appear.

---

I'll submit a PR with a test for this issue in a moment.

---

Related to https://github.com/squizlabs/PHP_CodeSniffer/issues/686.

**Comment by jrfnl** [comment]
@wvega Thanks for setting up that unit test, I'm looking into this now.

**Comment by jrfnl** [comment]
@wvega I've send you a PR - https://github.com/wvega/PHP_CodeSniffer/pull/1 - which will update PR #2392 when merged, which should fix this issue.

It would be great if you could have a look and test the PR (and merge it if you think the proposed solution is acceptable).

**Comment by wvega** [comment]
@jrfnl thank you for checking this issue so quickly. I just merged your PR. The solution looks great and fixes the problem I was having on my project.

**Comment by jrfnl** [comment]
@wvega It's something I'd come across a few times myself, but I hadn't investigated to determine the cause yet. Your clear explanation of the issue made it an easy one to (finally) solve now ;-)

**Comment by aik099** [comment]
While detecting (via debugging PHP_CodeSniffer) what `exclude-pattern` to use I've also noticed, that `\PHP_CodeSniffer\Filters\Filter::shouldIgnorePath` is processing combined array of ruleset + per-sniff exclude patterns and actually tries to compare sniff name against path. This doesn't do any bad for now, but if you have lots of per-sniff exclusion rules, then it would slow down file checking process.

Thanks for finding that out.

**Comment by peternolland** [comment]
fixed in PR #2392
